firmware/library/L3_HAL/button_test.cpp

```diff
@@ -0,0 +1,84 @@
+#include "L0_LowLevel/LPC40xx.h"
+#include "L3_HAL/button.hpp"
+#include "L5_Testing/testing_frameworks.hpp"
+
+TEST_CASE("Testing Button", "[button]")
+{
+  // Simulated local version of LPC_IOCON
+  // manipulation by side effect of Pin method calls
+  LPC_IOCON_TypeDef local_iocon;
+  // Substitute the memory mapped LPC_IOCON
+  // with the local_iocon test struture
+  memset(&local_iocon, 0, sizeof(local_iocon));
+
+  // Redirects manipulation to the 'local_iocon'
+  Pin::pin_map = reinterpret_cast<Pin::PinMap_t *>(&local_iocon);
+  LPC_GPIO_TypeDef local_gpio_port[2];
+  memset(&local_gpio_port, 0, sizeof(local_gpio_port));
+
+  // Only Port 1 is tested in this test.
+  Gpio::gpio_port[1] = &local_gpio_port[1];
+  // Onboard button of Port 1 and Pin 9 will be used for this test
+  Button test_subject(1, 9);
+
+  SECTION("Initialize")
+  {
+    constexpr uint8_t kInputSet     = 0b0;
+    constexpr uint8_t kPin9         = 9;
+    constexpr uint8_t kModePosition = 3;
+    constexpr uint32_t kMask        = 0b11 << kModePosition;
+
+    constexpr uint32_t kExpectedForPullDown =
+        static_cast<uint8_t>(PinInterface::Mode::kPullDown) << kModePosition;
+
+    test_subject.Initialize();
+
+    CHECK(kExpectedForPullDown == (local_iocon.P1_9 & kMask));
+    CHECK(((local_gpio_port[1].DIR >> kPin9) & 1) == kInputSet);
+  }
+  SECTION("Button Pressed and Button Released")
+  {
+    constexpr bool kExpectedResult = true;
+    constexpr uint8_t kPin9        = 9;
+
+    // Set P1_9 for button read == true
+    local_gpio_port[1].PIN |= (1 << kPin9);
+    // Simulate button being pressed.
+    test_subject.Released(true);
```

Comment:
... So I just realized how confusing this looks to pass a reset variable to the `Released` and `Pressed`, which is my bad, since I suggested it. I think having a `ResetState()`  function would be a cleaner alternative to this, so it pretty obvious what its doing.